fido/src/main/java/com/yubico/yubikit/fido/client/clientdata/JsonClientData.java

```diff
@@ -0,0 +1,265 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import javax.annotation.Nullable;
+
+/**
+ * Internal immutable representation of WebAuthn {@code clientDataJSON}.
+ *
+ * <p>Provides both the raw JSON bytes and their SHA-256 hash. Used when full {@code clientDataJSON}
+ * must be supplied (e.g. for attestation) rather than only its hash.
+ *
+ * <p>Construction either wraps provided raw bytes or builds deterministic JSON from field inputs:
+ *
+ * <ul>
+ *   <li>Standard fields: type, challenge (base64url), origin, crossOrigin, optional topOrigin.
+ *   <li>Extra parameters: filtered to exclude reserved keys and serialized in lexicographic key
+ *       order at the top level.
+ *   <li>Nested structures: values of type {@code Map} and {@code Iterable} (e.g. {@code List},
+ *       {@code Set}) are recursively serialized into JSON (not via {@code toString()}).
+ *   <li>Nested {@code Map} keys are emitted in the iteration order of the provided map (not
+ *       re-sorted). Use {@code LinkedHashMap} if deterministic ordering is required.
+ *   <li>Reserved key filtering applies only to the top-level extras; nested maps are not filtered.
+ *   <li>Binary ({@code byte[]}) values at any depth are base64url encoded as JSON strings.
+ *   <li>Other object types (non-primitive, non-collection, non-byte[]) are serialized via {@code
+ *       String.valueOf(value)} as JSON strings.
+ * </ul>
+ *
+ * <p>Reserved keys ignored in top-level extras: {@code type}, {@code challenge}, {@code origin},
+ * {@code crossOrigin}, {@code topOrigin}.
+ *
+ * <p>This class is package-private; external users interact via {@link ClientDataProvider}
+ * factories.
+ *
+ * @see ClientDataProvider
+ */
+final class JsonClientData implements ClientDataProvider {
+  private static final Set<String> RESERVED =
+      Collections.unmodifiableSet(
+          new HashSet<>(Arrays.asList("type", "challenge", "origin", "crossOrigin", "topOrigin")));
+
+  private final byte[] raw;
+  private final byte[] hash;
+
+  JsonClientData(byte[] raw) {
+    this.raw = raw.clone();
+    this.hash = Utils.hash(this.raw);
+  }
+
+  static JsonClientData createFromFields(
+      ClientDataType type,
+      byte[] challenge,
+      String origin,
+      boolean crossOrigin,
+      @Nullable String topOrigin,
+      @Nullable Map<String, ?> extraParameters) {
+    byte[] json =
+        buildClientDataJson(
+            type.jsonValue(), challenge, origin, crossOrigin, topOrigin, extraParameters);
+    return new JsonClientData(json);
+  }
+
+  @Override
+  public byte[] getHash() {
+    return hash.clone();
+  }
+
+  @Override
+  public byte[] getClientDataJson() {
+    return raw.clone();
+  }
+
+  @Override
+  public boolean hasClientDataJson() {
+    return true;
+  }
+
+  private static byte[] buildClientDataJson(
```

Comment:
Rather than flipping booleans and removing reserved keys from extras I would prefer if we threw IllegalArgumentExceptions. If the caller is passing illegal combinations, a direct error is probably more clear than silently adjusting things to make it valid.